compiler/rustc_typeck/src/expr_use_visitor.rs

```diff
@@ -265,6 +265,9 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
                                     if def.variants.len() > 1 {
                                         needs_to_be_read = true;
                                     }
+                                } else {
+                                    // If it is not ty::Adt, then it should be read
+                                    needs_to_be_read = true;
```

Comment:
I am wondering what cases might be missing here -- one silly example might be matching on a constant of type `()` or something like that, which is kind of a no-op.

Reply:
I am not too sure either. I am going to import all the match test cases that are here https://github.com/rust-lang/rust/tree/master/src/test/ui/match and then convert them into closure test cases and see if I found anything new